src/Ast.ml

```diff
@@ -273,6 +273,24 @@ module Structure_item : Module_item with type t = structure_item = struct
        |Pstr_extension _, Pstr_extension _ ->
           true
       | _ -> false )
+    | `Preserve, `Preserve -> (
+      match (itmI.pstr_desc, itmJ.pstr_desc) with
+      | Pstr_eval _, Pstr_eval _
+       |Pstr_primitive _, Pstr_primitive _
+       |(Pstr_type _ | Pstr_typext _), (Pstr_type _ | Pstr_typext _)
+       |Pstr_exception _, Pstr_exception _
+       |( (Pstr_module _ | Pstr_recmodule _ | Pstr_open _ | Pstr_include _)
+        , (Pstr_module _ | Pstr_recmodule _ | Pstr_open _ | Pstr_include _)
+        )
+       |Pstr_modtype _, Pstr_modtype _
+       |Pstr_class _, Pstr_class _
+       |Pstr_class_type _, Pstr_class_type _
+       |Pstr_attribute _, Pstr_attribute _
+       |Pstr_extension _, Pstr_extension _ ->
+          true
+      | Pstr_value (_, [{pvb_loc= x}]), Pstr_value (_, [{pvb_loc= y}]) ->
+          not (Source.empty_line_between x y)
```

Comment:
Couldn't you refactor this to avoid duplicating these many cases by combining the Compact and Preserve cases, and adding a check of `module_item_spacing` in the `Pstr_value` case? Maybe use `space_between` defined below to simplify?